### PR TITLE
Native support for byte array

### PR DESCRIPTION
**Pull request by dongche**
Native support for byte array:
* add native methods. Change the code from Diceros a little.
* interface method param is added output and its offset

@jerrychenhf , @dianfu

### CODE REVIEW

File: src/main/java/com/intel/chimera/cipher/Cipher.java [hunk 1]
```diff
@@ -76,6 +76,21 @@ int update(ByteBuffer inBuffer, ByteBuffer outBuffer)
...
+   *
+   * @param input the input byte array
+   * @param inputOffset the offset in input where the input starts
+   * @param inputLen the input length
+   * @param output the byte array for the result
+   * @param outputOffset the offset in output where the result is stored
+   * @return the number of bytes stored in output
+   * @throws ShortBufferException if there is insufficient space in the output byte array
+   */
+  int update(byte[] input, int inputOffset, int inputLen, byte[] output, int outputOffset)
```
Comment (jerrychenhf) [inline src/main/java/com/intel/chimera/cipher/Cipher.java]: Suggest to change line if there are two many parameters. For example:

int update(byte[] input, int inputOffset, int inputLen,
     byte[] output, int outputOffset)


The same for others if there are 5 or more parameters.

File: src/main/java/com/intel/chimera/cipher/OpensslNative.java [hunk 2]
```diff
@@ -31,8 +31,15 @@ public native static int update(long context, ByteBuffer input,
       int inputOffset, int inputLength, ByteBuffer output, int outputOffset,
       int maxOutputLength);
 
+  public native static int updateByteArray(long context, byte[] input,
```
Comment (jerrychenhf) [inline src/main/java/com/intel/chimera/cipher/OpensslNative.java]: Is it technical possible to use the same name "update" instead of "updateByteArray"?

**Comment by jerrychenhf** [comment]
Thanks @dongche for putting up the patch!
I leave a few of comments for minor issues on formats.

File: src/main/java/com/intel/chimera/cipher/Cipher.java (hunk 1 above)
Comment (dongche) [inline src/main/java/com/intel/chimera/cipher/Cipher.java]: Done

File: src/main/java/com/intel/chimera/cipher/OpensslNative.java (hunk 2 above)
Comment (dongche) [inline src/main/java/com/intel/chimera/cipher/OpensslNative.java]: Since cannot overload methds in C, we have to differ the methods name.

**Comment by dongche** [comment]
Thanks @jerrychenhf  for your review! Patch updated to refine the format.

**Comment by jerrychenhf** [comment]
Thanks @dongche  so much for quick addressing my comments. Good to me for the functional code part.
For test case part, only doFinal is tested, I would suggest also test update combine with the doFinal.

**Comment by dongche** [comment]
Oh, yes, testing update combined with doFinal is missed. Patch updated to add them.

**Comment by dongche** [comment]
One thing need to be mentioned is: In the test, when decrypt cipherText, the output length also have to be cipherText's length plus blockSize. Otherwise, AES/CBC/PKCS5Padding mode will get exception of not sufficient output buffer. See https://github.com/intel-hadoop/chimera/pull/80/files/e692c566be375d6bc1da1a6befb79c73550a927b#diff-ea08ae72ffde664de2d1effe064b9903R190

In theory it seems not be needed to plus block size when decrypt. I will take a look if it is ok or can be improved.

Anyway, this does not block this PR to proceed.

**Comment by jerrychenhf** [comment]
Agree that for the output buffer size estimation and checking, we can put it to another PR. Let's check that whether openssl has a method to return output size giving the input len.

**Comment by jerrychenhf** [comment]
Thanks @dongche for updating the patch. Very close now.
 
The data length and the data pieces used to call update are all multiple of AES block size. It is very important to have the test cases also test cases that not multiple block size for both dataLen and bufferLen.

 171   +    int dataLen = 10 * 1024;  
 177   +    int bufferLen = 2 * 1024 - 128;

**Comment by dongche** [comment]
@jerrychenhf thanks for the comments of testing not aligned data. Patch updated to refine the test cases to cover more situations.

**Comment by jerrychenhf** [comment]
Looks good to go.

**Pull request merged**